### PR TITLE
User roles

### PR DESCRIPTION
**Pull request by ahmadkurdo**
**Getting users information:**
-  Send a GET request to `http://127.0.0.1:8000/users/users/` this will return the user's information including a field called role. The role can be either user, garage_owner, or admin. Use the role field to view the correct view to the user.

**NOTE: In order to access all endpoints below you need to be logged in as admin otherwise you will get the following response:  `{'message': unauthorized access'}`**

**Making a garage_owner:**

- Send a POST request to `http://127.0.0.1:8000/users/admin/create_garage_owner/` with the following JSON: `{"first_name":"","last_name":"","iban":"","address":"","house_number":"","zip_code":"","email":"", "password":""}`

**Making an admin:**
- Send a POST request to `http://127.0.0.1:8000/users/admin/` with the following JSON: `{"first_name":"","last_name":"","iban":"","address":"","house_number":"","zip_code":"","email":"", "password":""}`

**Getting all users:**

-  Send a GET request to `http://127.0.0.1:8000/users/admin/`

### CODE REVIEW

**Comment by buzzzlightyear** [comment]
**Review**
1. Sending a GET request to `http://127.0.0.1:8000/users/users/` doesn't work, did you mean: `http://127.0.0.1:8000/users/userinfo/`?
2. When sending a request to `http://127.0.0.1:8000/users/userinfo/` I get the following response:
```json
{
    "first_name": "",
    "last_name": "",
    "iban": "INGB999999999",
    "address": "Wijnhaven",
    "house_number": "107",
    "zip_code": "3333PP",
    "user": 5,
    "role": ""
}
```
- The value of `first_name` and `last_name` when empty should return `null` , this is cleaner.
- The value of `role` shouldn't be a varchar. I explained the problems this would have, yet you still did it.. Please change it to an integer with one<>many relationship with a table called something like `user_roles` (ID | Name). **Then** when preparing this json response, you can look up the role in the `user_role` table and return it as a string. So in the `auth_user` the `role` field is an `int` and when returning the json response you look up the field and return the role as a string.
- I miss the e-mail!!
- `user` doesn't say anything, you should change it to `user_id`
- Remove `address` and add `street_name`, `city` and `house_number_addition`. This is a bit more professional.

After you've implemented those changes, request a review and I'll review it again.

**Comment by ahmadkurdo** [comment]
> **Review**
> 
> 1. Sending a GET request to `http://127.0.0.1:8000/users/users/` doesn't work, did you mean: `http://127.0.0.1:8000/users/userinfo/`?
> 2. When sending a request to `http://127.0.0.1:8000/users/userinfo/` I get the following response:
> 
> ```json
> {
>     "first_name": "",
>     "last_name": "",
>     "iban": "INGB999999999",
>     "address": "Wijnhaven",
>     "house_number": "107",
>     "zip_code": "3333PP",
>     "user": 5,
>     "role": ""
> }
> ```
> 
> * The value of `first_name` and `last_name` when empty should return `null` , this is cleaner.
> * The value of `role` shouldn't be a varchar. I explained the problems this would have, yet you still did it.. Please change it to an integer with one<>many relationship with a table called something like `user_roles` (ID | Name). **Then** when preparing this json response, you can look up the role in the `user_role` table and return it as a string. So in the `auth_user` the `role` field is an `int` and when returning the json response you look up the field and return the role as a string.
> * I miss the e-mail!!
> * `user` doesn't say anything, you should change it to `user_id`
> * Remove `address` and add `street_name`, `city` and `house_number_addition`. This is a bit more professional.
> 
> After you've implemented those changes, request a review and I'll review it again.

- Django returns a empty string for every null field this is just the Django way.

- I don't see a reason for doing this.I have coded the endpoint in a way that we won't .I will keep it as it is.

- It used to be user_id you told me to change it to user I'am not going to change it every again.

- I'll implement email address

- Changing this will not be that easy as I have to modify the database and the users info endpoint this will cause a lot of bugs. These fields were mentioned in the ERD I believe these changes should have been mentioned in the beginning of the project.

File: api/users/views.py [hunk 1]
```diff
@@ -99,48 +98,58 @@ def user_info(request):
...
-            return JsonResponse({'message': 'admin created'})
+            user = admin.createUser()
+            request_handel.add_to_dictionary("user", user.id)
+            serializer_handler.createInstance(
+            serializer=CustomerSerializerForAdmins(data=request_handel.get_dictionary()),
+            newObjectsMessage='newCustomer', message='Customer created')
+            return JsonResponse(serializer_handler.data, status=serializer_handler.status)
     else:
-        return JsonResponse({"message":"Unauthorized access"})
+        return JsonResponse({"message":"Unauthorized access"}, status=403)
```
Comment (buzzzlightyear) [inline api/users/views.py:152]: These 2 view functions are executing a lot of the same logic. It is simply code duplication to the fullest. When coding, please keep in mind, if you find yourself duplicating code: then there is always a way to refactor it and make it cleaner. So no code repetition.

You could easily remove 20 -50 lines of code. My suggestion is to create 1 view for adding a new user. Which user? That must come to knowledge when reading out the request object.

File: api/users/urls.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
     path('car/', views.user_car),
     path('edit_car/', views.edit_car),
     path('userinfo/', views.user_info),
-    path('admin/', views.Admin),
-    path('admin/create_admin/', views.create_admin),
+    path('admin/', views.admin),
+    path('admin/create_garage_owner/', views.create_garage_owner),
```
Comment (buzzzlightyear) [inline api/users/urls.py:13]: See comment below, you could just make one path for making any user of any type.

File: api/users/models.py [hunk 3]
```diff
@@ -12,15 +12,28 @@ def getOwner(self):
...
 class CustomerUser(models.Model):
+    first_name = models.CharField(max_length=100)
+    last_name = models.CharField(max_length=100)
     iban = models.CharField(max_length=100)
     address = models.CharField(max_length=100)
     house_number = models.CharField(max_length=3)
     zip_code = models.CharField(max_length=4)
-    is_garage_owner = models.BooleanField(default=False)
     user = models.OneToOneField(User,on_delete=models.CASCADE, primary_key=True)
+    role = models.CharField(max_length=30, default='user')
```
Comment (buzzzlightyear) [inline api/users/models.py:22]: No varchar. int with one<>many to table `user_roles`.

File: api/users/views.py [hunk 4]
```diff
@@ -99,48 +98,58 @@ def user_info(request):
...
             customer_response = {}
             for customer in customers:
                 serializer = CustomerSerializer(customer)
-                customer_response[serializer.data['user_id']] = serializer.data
+                print(serializer.data)
+                customer_response[serializer.data['user']] = serializer.data
             return JsonResponse({'customers': customer_response}, status=201)
     else:
-        return JsonResponse({"message":"Unauthorized access"})
+        return JsonResponse({"message":"Unauthorized access"}, status=403)
```
Comment (buzzzlightyear) [inline api/users/views.py:129]: Dutch response!

File: api/users/views.py [hunk 5]
```diff
@@ -99,48 +98,58 @@ def user_info(request):
...
-    if user.is_superuser:
+    if user.is_superuser or customer.is_admin():
         if request.method == 'POST':
             user = admin.createUser()
-            request_handel.add_to_dictionary("user_id", user.id)
+            request_handel.add_to_dictionary("user", user.id)
             serializer_handler.createInstance(
-            serializer=CustomerSerializer(data=request_handel.get_dictionary()),
+            serializer=CustomerSerializerForAdmins(data=request_handel.get_dictionary()),
             newObjectsMessage='newCustomer', message='Customer created')
```
Comment (buzzzlightyear) [inline api/users/views.py:116]: Dutch response!

**Pull request closed without merging**